src/nxdoc/nuxeo-server/data-store/indexing-and-query/elasticsearch-hints-cheat-sheet/how-to-make-elasticsearch-hints-extension-point.md

```diff
@@ -0,0 +1,84 @@
+---
+title: 'HOWTO: Expose Elasticsearch Hints as Extension Points'
+description: 'Learn how to contribute or override an existing Elasticsearch Hint'
+review:
+    comment: ''
+    date: '2019-12-09'
+    status: ok
+labels:
+    - addon
+    - tutorial
+    - howto
+toc: true
+tree_item_index: 100
+---
+
+##  Concept
+
+Elasticsearch hints can be exposed as extension point: it is therefore possible to create a **custom hint** in order to use all the possible options offered by the Elasticsearch operators. It provides a good way no to be restricted to the Nuxeo default exposed operators, listed in the [Nuxeo NXQL Hints page]({{page space='nxdoc' page='nxql'}}#elasticsearch-nxql-hints). The idea is to replace the hard coded NXQL hint with contribution system.
+
+## Elasticsearch hint contribution
+
+To contribute or override an existing Elasticsearch hint, you should provide a name for your Elasticsearch hint and an implementation of `org.nuxeo.elasticsearch.api.ESHintQueryBuilder`.
+Each hint would have **a name and a class** that implement the same java interface with one method that takes the three parameters (field, analyzer and value) and returns the resulting ES query object. The parsing process of query is not impacted. Only the way of how we do the correspondence between the ES hint (NXQL operator) and the ES query.
+
+##  Use cases
+
+### Fuzzy searches
+
+The [Elasticsearch fuzzy search operator](https://www.elastic.co/guide/en/elasticsearch/reference/6.5/query-dsl-fuzzy-query.html) contains a list of [parameters](https://www.elastic.co/guide/en/elasticsearch/reference/6.5/query-dsl-fuzzy-query.html#_parameters_7). By  default, Nuxeo doesn't expose them directly in NXQL queries: only the default one are taken into account. For example, it only works on a single term (not analysed), which is a limited use case; a match operator with fuzziness will be more interesting.
+
+It can be solved by creating a custom hint, called for example **myFuzzy** which adapts the parameters to our needs. Then, it can be used in any NXQL queries like:
+
+```
+SELECT * FROM Document WHERE /*+ES: OPERATOR(myFuzzy) */ dc:title = 'foo'
+```
+
+"fuzzy" is our case and a class which must implement the interface below  
+
+```
+public interface ESHintQueryBuilder {
+ QueryBuilder make(EsHint hint, String fieldName, Object value);
+}
+
+public class FuzzyESHintQueryBuilder implements ESHintQueryBuilder {
+
+ @Override
+ public QueryBuilder make(EsHint hint, String fieldName, Object value) {
+ return QueryBuilders.fuzzyQuery(fieldName, (String) value);
+ }
+}
+```
+
+The idea is to move each case of our switch statement to a new implementation of **ESHintQueryBuilder**. Each contribution lookup is done by his name.
```

Comment:
i am wondering if we should talk about the fact that we had in the old way a `switch` statement and now we change it. Perhaps just say:

Now the new way to expose the the ES hints is doing by contribution as below: ......